### ISSUE TITLE
[FEAT] Triggers and Events

### ISSUE THREAD

**Issue by dthv** [Kruptein/PlanarAlly#983]
Sometimes a DM can overlook actions of players that would normally trigger an event (e.g. a booby-trap triggered by a trip-wire).

Thus, it would be nice, if Shapes could be equipped with triggers that trigger certain events.

Possible Triggers include:
- movement: Token moves on Shape, activating trigger
- sight
  - Token sees Shape, activating Shape's trigger (consider e.g. Medusae)
  - Shape sees Token, activating Shape's trigger
- activate (triggered by Event, see below)

Possible Events include:
- message/popup
 - displayed either to a) DM-only, b) DM and player only, c) public.
 - not sure whether popup or disappearing message
- entanglement, locking (to pause game for event resolution)
  - triggering Token
  - the whole group (thus, pausing game completely until resolution)
- modify (e.g. lock/unlock doors etc.)
- dice roll, rolling pre-defined combination of dice (either DM-only or public), adding result to message (if any)
- chain, activating trigger of another Shape (should probably not be able to be linked to activation, to prevent loops in the event chain; this can probably be circumvented, though)

Triggers should be able to deactivate and still be part of a Shape's information.
Events should be dismissable with one click, reversing lock effects.
It should be possible to trigger more than one event with only one trigger (e.g. display a message and lock triggering Token and roll dice at only one Token moving on Shape.

The *activate" trigger and the *modify* and *chain* events are prone to add a hell of complexity. It can be a mighty tool to automatically change the entire map in an instance, supporting fast and immersive play. Yet, errors in the setup *can*, in principle, create loops and break the game. Thus, a "deactivate all Triggers" should be a action that can be performed on a Campaign from the Dashboard. It will help DMs to debug their Games.

**Comment by develroo** [comment]
Yes, I can see this being useful. Essentially, you are taking the teleport logic and using it to trigger another event. That could get complex fast if you don't have defined events to target. To my mind the easiest would be:

* Flipping a token variant from one image to another
* Toggling tokens that were invisible to be visible (like a pit or fire trap)
* Toggling Annotations from private to public 
* Toggling a player's movement to freeze them in place.

Essentially, the idea is to keep such event chains to one or other existing functions, which should save on complexity. 

But I like the idea, as currently I am manually flipping tokens and things.
